### PR TITLE
fix:fix server address error when running in k8s

### PR DESCRIPTION
**Pull request by golangaccount**
when goc server running  in k8s
client register uri while be transform to is proxy address
in fact,there is no need to transform

add address uri check,make sure address parm is a complete uri

### CODE REVIEW

**Comment by CarlJi** [comment]
@golangaccount Thanks for your contribution.

The reason of this transform is try to find the proper IP especially in the local multi-interfaces scenario. And yes it's not elegant for the kubernetes scenario. To enhance it, we have solved this drawback in the goc v2 version by using web-socket protocol and welcome to have a try. https://github.com/qiniu/goc/tree/v2

For your PR, i am thinking that it's better to provide a switch in `goc server` to let user choose whether to do this transform or not than changing old behavior. since the default behavior should be compatibility with previous versions。

**Comment by golangaccount** [comment]
cureent, i use goc v1.x in project,so i think has necessary to merge v1.x.   yea,use this code,the nat network will not work well,i will add a parm in goc server to switch two conditions

**Comment by golangaccount** [comment]
@CarlJi  i edit the code，add a parm to control the network，if not set，it runing default

File: pkg/cover/server.go [hunk 1]
```diff
@@ -38,6 +38,7 @@ const LogFile = "goc.log"
 
 type server struct {
 	PersistenceFile string
+	NetworkType     string //solve different network type(1.regist client can direct access ,2.regist client under a proxy network or nat network、same network ect)
```
Comment (CarlJi) [inline pkg/cover/server.go:41]: The concept  of `NetworkType` seems too  big, actually the goal here is only to check whether to do ip revise during registering。so i prefer:
```go
type server struct{
   ...
   IPRevise bool // whether to do ip revise during registering
```

File: pkg/cover/server.go [hunk 2]
```diff
@@ -123,24 +125,53 @@ func (s *server) registerService(c *gin.Context) {
...
+			return
+		}
 
-	realIP := c.ClientIP()
-	// only for IPV4
-	// refer: https://github.com/qiniu/goc/issues/177
-	if net.ParseIP(realIP).To4() != nil && host != realIP {
-		log.Printf("the registered host %s of service %s is different with the real one %s, here we choose the real one", service.Name, host, realIP)
-		service.Address = fmt.Sprintf("http://%s:%s", realIP, port)
+		realIP := c.ClientIP()
```
Comment (CarlJi) [inline pkg/cover/server.go:168]: I am thinking that no need to change the `register api`, we can just use `goc sever` to control this behave. like, Launch go server with `goc server --ip_revise=false`
And change here's logic as
```go 
	// only do ip revise as needed.
	if s.IPRevise {
		realIP := c.ClientIP()
		// only for IPV4
		// refer: https://github.com/qiniu/goc/issues/177
		if net.ParseIP(realIP).To4() != nil && host != realIP {
			log.Printf("the registered host %s of service %s is different with the real one %s, here we choose the real one", service.Name, host, realIP)
			service.Address = fmt.Sprintf("http://%s:%s", realIP, port)
		}
	}
```

File: pkg/cover/server.go (hunk 1 above)
Comment (CarlJi) [reply pkg/cover/server.go:41]: `IPRevise` can be passed when goc server lauching like `goc server --ip_revise=false`.
Offically by default, `IPRevise` should be true to keep consistent with previous versions.

**Review by CarlJi — CHANGES_REQUESTED** [review]

**Comment by CarlJi** [comment]
Hi @golangaccount , if i understood your scenario correctly, switch off the ip revise during registering should meet your requirements。 so no need to change the logic of 'registry API', since it will affect the old behaviors。

File: pkg/cover/server.go [hunk 3]
```diff
@@ -123,24 +125,53 @@ func (s *server) registerService(c *gin.Context) {
...
+	}
 	u, err := url.Parse(service.Address)
 	if err != nil {
 		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
 		return
 	}
-	host, port, err := net.SplitHostPort(u.Host)
-	if err != nil {
-		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+	if u.Scheme != "https" && u.Scheme != "http" {
```
Comment (CarlJi) [inline pkg/cover/server.go:136]: why do we need this check？

**Comment by golangaccount** [comment]
@CarlJi use bool can solve my problem，but there also has some question. if goc server is running, some client already regist,now i want to regist a new client with no revise,so i must set the value, use bool canot make sure the value is user set.

**Comment by CarlJi** [comment]
while, if we want to let user be able to control this ability during registering from client side, i actually expect a new PR since it will also need to be changed for the `goc build/install` logic.

for your questions, i suggest to deploy multi goc servers. In my original design concepts, goc server should be lightweight and close to the services under tests。(PS: maybe it should be called `goc proxy` ... )  :sweat_smile:

**Comment by golangaccount** [comment]
@CarlJi i change the parm,use ip_revise in goc server. i still retain request parm (iprevise) to change the server running style

**Comment by CarlJi** [comment]
Hi @golangaccount , i am curious do you just use `goc server` alone？and how do you do instrumentation for your services without `goc build/install` ?

**Comment by golangaccount** [comment]
@CarlJi  i also use goc build/install,  goc server --ip_revise=false can solve my k8s question(other code no need change). 

> when goc server running in k8s client register uri while be transform to is proxy address in fact,there is no need to transform
> 
> add address uri check,make sure address parm is a complete uri

**Pull request merged**